src/main/java/org/apache/flink/benchmark/SinkUpsertMaterializerBenchmark.java

```diff
@@ -0,0 +1,298 @@
+SinkUpsertMaterializerBenchmark.run            true            100              10                  100         ROCKSDB            V1  thrpt        209.375          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100             100                  100            HEAP            V1  thrpt       2131.445          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100             100                  100         ROCKSDB            V1  thrpt         78.209          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100             200                  100            HEAP            V1  thrpt        652.936          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100             200                  100         ROCKSDB            V1  thrpt         29.674          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100            1000                  100            HEAP            V1  thrpt        118.567          ops/ms
+SinkUpsertMaterializerBenchmark.run            true            100            1000                  100         ROCKSDB            V1  thrpt          6.426          ops/ms
+ * }</pre>
+ */
+@OperationsPerInvocation(value = SinkUpsertMaterializerBenchmark.RECORDS_PER_INVOCATION)
+@SuppressWarnings("ConstantValue")
+public class SinkUpsertMaterializerBenchmark extends BenchmarkBase {
+
+    private static final int STREAM_KEY = 0;
+
+    public static void main(String[] args) throws RunnerException {
+        new Runner(new OptionsBuilder()
+                .verbosity(VerboseMode.NORMAL)
+                // speedup
+//                .warmupIterations(1)
+//                .measurementIterations(1)
+//                .forks(1)
+//                .warmupTime(TimeValue.milliseconds(100))
+//                .measurementTime(TimeValue.seconds(1))
+                .include(".*" + SinkUpsertMaterializerBenchmark.class.getCanonicalName() + ".*")
+                .build()).run();
+    }
+
+    @Benchmark
+    public void run(SumBmState state) throws Exception {
+        for (long record = 0; record < state.numRecordsTotal; record++) {
+            state.harness.processElement(insertRecord(record, STREAM_KEY, state.payload));
+            if (state.shouldRetract(record)) {
+                state.harness.processElement(
+                        deleteRecord(record - state.retractOffset, STREAM_KEY, state.payload));
+            }
+        }
+    }
+
+    protected static final int RECORDS_PER_INVOCATION = 10_000;
+
+    @State(Scope.Thread)
+    public static class SumBmState {
+
+        @Param({"false", "true"}) public boolean hasUpsertKey;
+
+        @Param({"HEAP", "ROCKSDB"}) public SumStateBackend stateBackend;
+
+        public int numRecordsTotal;
+
+        // larger payload amplifies any inefficiencies but slows down the benchmark; mostly affects rocksdb
+        @Param("100")
+        public int payloadSize;
```

Comment:
For micro benchmark purposes, like optimising code, guarding against regressions, the lower the payload, the better, as you get more contrasting effects of any performance related changes - optimisations/changes to the related code will be more clearly visible instead of being watered down by bytes copy.

So I would shrink it down, maybe to 10 or 20.

Reply:
I think both cases can be problematic; I'll change to a smaller value and think about keeping 100 (or increasing it).